src/coreclr/src/vm/class.cpp

```diff
@@ -1184,107 +1184,112 @@ void ClassLoader::ValidateMethodsWithCovariantReturnTypes(MethodTable* pMT)
         return;
 
     // Step 1: Validate compatibility of return types on overriding methods
-
-    for (WORD i = 0; i < pParentMT->GetNumVirtuals(); i++)
+    if (pMT->GetClass()->HasCovariantOverride())
     {
-        MethodDesc* pMD = pMT->GetMethodDescForSlot(i);
-        MethodDesc* pParentMD = pParentMT->GetMethodDescForSlot(i);
+        for (WORD i = 0; i < pParentMT->GetNumVirtuals(); i++)
```

Comment:
Would `MethodTable::MethodIterator` made this faster? (Not sure whether it would - just asking.)

Reply:
Its hard to say MethodIterator is MethodData based. Working with MethodData objects is one of the most expensive things that the type loader does (it regularly shows up in profiles), but this code is also slow, as resolving back to the MethodDesc is enormously slow. That's why I took the alternative approach of simply not running this code unless we were in a situation where it was needed.

Also, the semantics of MethodData are somewhat confusing. I'm not sure that translating to use MethodData would be trivial.